Migrate credits to VGUI frame

braem opened the pull request:
Closes #823

Moves the credits logic from an env in credits map to a VGUI frame. Also updated the credits to the best of my ability (will bring it up in the meeting if people want their name changed & such).

Couple things:
- Couldn't just set the background color to black, had to throw a couple getter/setters in `Frame.cpp` that get/set the in and out of focus bg colors.
- Left out the call to game stats, `Event_Credits`, as that's in the server. There are a couple calls in that function that are also in the server, not sure what to do there.
- The input credits and logo methods aren't used at all. Will we need these later or are they safe to remove?
- Had to play music with `enginesound` rather than `surface()`, as it seems there's no way to stop a sound using it.
- Removed all references I could find to the credits map, removed some HUD hiding occurring during credits rolling, and removed `"disconnect"` from the list of allowed client commands in the game rules.
- Tested before and after changing resolutions

Once this is approved I'd like to make one last commit moving and renaming the file. Left it out for now to avoid pain in the butt rebases.

### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

braem commented:
Gunna guess the failing Linux builds are due to me not capitalizing the `engine/IEngineSound.h` include correctly.

braem commented:
> 
> 
> Gunna guess the failing Linux builds are due to me not capitalizing the `engine/IEngineSound.h` include correctly.

🙌🐢

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -137,8 +120,50 @@ class CHudCredits : public CHudElement, public Panel
+    SetMaximizeButtonVisible(false);
+    SetCloseButtonVisible(false);
+
+    int iWidth, iTall;
+    surface()->GetScreenSize(iWidth, iTall);
+    SetBounds(0, 0, iWidth, iTall);
```
Override `PerformLayout` and do this there, as this will help with resolution changes.

braem replied:
Originally had it there, though the credits preparation, which calls `GetWide()` and `GetTall()` happens before the `PerformLayout` causing some buggy behavior. Might be fine when I create that separate resource file as the wide and tall will be loaded in the constructor.

Gocnak replied:
You can do it in both. And regardless of implementation, ensure the panel is full screen upon resolution changes.

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -137,8 +120,50 @@ class CHudCredits : public CHudElement, public Panel
+    surface()->GetScreenSize(iWidth, iTall);
+    SetBounds(0, 0, iWidth, iTall);
+
+    sv_unlockedchapters.SetValue("15");
```
Since [this is set to 99 by default](https://github.com/momentum-mod/game/blob/develop/mp/src/game/server/gameinterface.cpp#L2004) we can remove this ref, as it's old behavior from HL2 I believe.

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -217,7 +240,10 @@ void CHudCredits::ReadNames(KeyValues *pKeyValue)
         V_strcpy_safe(Credits.szCreditName, pKVNames->GetName());
-        V_strcpy_safe(Credits.szFontName, pKeyValue->GetString(Credits.szCreditName, "Default"));
+
+        const auto hScheme = scheme()->GetScheme("ClientScheme");
+        const auto pScheme = scheme()->GetIScheme(hScheme);
```
Move this out of the while loop

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -26,7 +23,7 @@ using namespace vgui;
 {
     char szCreditName[256];
-    char szFontName[256];
+    HFont font;
```
Nice :)

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -103,7 +85,8 @@ class CHudCredits : public CHudElement, public Panel
     void PrepareLine(HFont hFont, char const *pchLine);
 
     CPanelAnimationVar(vgui::HFont, m_hTextFont, "TextFont", "Default");
-    CPanelAnimationVar(Color, m_TextColor, "TextColor", "FgColor");
+    CPanelAnimationVar(Color, m_TextColor, "TextColor", "White");
+    CPanelAnimationVar(Color, m_bgColor, "BgColor", "Black");
```
Since this is no longer a HUD element, where would these get set? You should do `LoadControlSettings("resource/ui/CreditsDialog.res");` in the constructor, with the panel itself being the only entry. You'd then be able to set these colors in there.

Gocnak replied:
I also would recommend setting the scheme to something in the constructor, and making sure to use it throughout the rest of this file. You can create a CreditsScheme.res file that just bases off of the source scheme, look at ConsoleScheme.res for more info on that.

braem replied:
Was thinking this actually though felt it was weird to expose credits stuff to res files. I suppose the names themselves are exposed in `scripts/credits.txt` so that makes sense.

Gocnak replied:
Well the goal is to allow people to completely customize the UI if they want, which this would significantly help them do so.

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -137,8 +120,50 @@ class CHudCredits : public CHudElement, public Panel
+//-----------------------------------------------------------------------------
+// Purpose: Constructor
+//-----------------------------------------------------------------------------
+CHudCredits::CHudCredits() : BaseClass(nullptr, "HudCredits"), sv_unlockedchapters("sv_unlockedchapters")
```
Can rename this panel to remove the "HUD" references, as it is no longer a HUD element. I would suggest the name `CreditsDialog`.

Gocnak commented on mp/src/public/vgui_controls/Frame.h:
```diff
@@ -133,6 +133,11 @@ class Frame : public EditablePanel
 	// Temporarily enables or disables the fade effect rather than zeroing the fade times as done in DisableFadeEffect
 	void SetFadeEffectDisableOverride( bool disabled );
 
+    void SetInFocusBgColor(Color col);
+    void SetOutOfFocusBgColor(Color col);
+    Color GetInFocusBgColor();
+    Color GetOutOfFocusBgColor();
```
Since these are basic getters and setters, they can be defined in the header:
```suggestion
    void SetInFocusBgColor(Color col) { m_InFocusBgColor = col; }
    void SetOutOfFocusBgColor(Color col) { m_OutOfFocusBgColor = col; }
    Color GetInFocusBgColor() const { return m_InFocusBgColor; }
    Color GetOutOfFocusBgColor() const { return m_OutOfFocusBgColor; }
```

braem replied:
Actually won't even need these I think, as they are able to be overwritten in a resource control file.

Gocnak commented:
Regarding:
> Left out the call to game stats, Event_Credits, as that's in the server. There are a couple calls in that function that are also in the server, not sure what to do there.

Used for achievements in HL2 or something I think. When we figure out game stats we will figure out what we want to do with that.

>The input credits and logo methods aren't used at all. Will we need these later or are they safe to remove?

Yep. I don't know of any map with custom credits, they just use game_text iirc.

braem commented:
Implemented the requested changes:
- Created `CreditsScheme.res` and `CreditsDialog.res`, loaded them in constructor
- Removed a ton of unusued credits code
- Renamed the class, fonts, methods, etc. to remove mentions of "hud" and "outro"
- Removed the extra getters/setters in `Frame.cpp` as they are overridden in `CreditsScheme.res`

Gocnak commented on mp/src/game/server/EnvMessage.cpp:
```diff
@@ -145,134 +144,3 @@ void CMessage::Use( CBaseEntity *pActivator, CBaseEntity *pCaller, USE_TYPE useT
-	float		m_flLogoLength;
-};
-
-LINK_ENTITY_TO_CLASS( env_credits, CCredits );
```
You should also remove this entity from the base FGD

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -319,7 +256,7 @@ void CHudCredits::DrawOutroCreditsName(void)
                         {
                             pCredit->bActive = false;
-                            engine->ClientCmd("creditsdone");
+                            CloseModal();
```
You could probably also early return here, too. Saves rendering the panel while it closes?

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -165,135 +140,97 @@ void CHudCredits::PrepareCredits(const char *pKeyName)
-}
-
-void CHudCredits::DrawOutroCreditsName(void)
+void CreditsDialog::DrawCreditsName(void)
```
Remove unnecessary `(void)` parenthesis

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -113,33 +70,50 @@ class CHudCredits : public CHudElement, public Panel
 
-    float m_flX;
-    float m_flY;
+    const auto settingsScheme = scheme()->LoadSchemeFromFile("resource/CreditsScheme.res", "CreditsScheme");
+    SetScheme(settingsScheme ? settingsScheme : scheme()->GetDefaultScheme());
```
Panels without a scheme default to the default scheme, so this can just be an if:
```suggestion
    const auto settingsScheme = scheme()->LoadSchemeFromFile("resource/CreditsScheme.res", "CreditsScheme");
    if (settingsScheme)
    {
        SetScheme(settingsScheme);
    }
```

braem replied:
I copied this from the settings dialog, didn't even rename it, woops. I did the requested change and changed that variable name to `creditsScheme`.

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -113,33 +70,50 @@ class CHudCredits : public CHudElement, public Panel
-    float m_flFadeOutTime;
-    float m_flNextStartTime;
-    float m_flPauseBetweenWaves;
+//-----------------------------------------------------------------------------
+// Purpose: Constructor
+//-----------------------------------------------------------------------------
```
Remove empty / duh comments

Gocnak commented on mp/src/game/shared/momentum/mom_gamerules.cpp:
```diff
@@ -292,9 +291,8 @@ void CMomentumGameRules::PlayerSpawn(CBasePlayer *pPlayer)
     if (pPlayer)
     {
         ConVarRef map("host_map");
```
Not needed anymore

Gocnak commented on mp/src/game/client/hl2/hud_credits.cpp:
```diff
@@ -647,90 +348,24 @@ void CHudCredits::PrepareOutroCredits(void)
-}
-
-void CHudCredits::MsgFunc_CreditsMsg(bf_read &msg)
+CON_COMMAND_F(mom_credits_show, "Shows the credits.\n",
+              FCVAR_CLIENTCMD_CAN_EXECUTE | FCVAR_SERVER_CAN_EXECUTE | FCVAR_HIDDEN)
```
Can be one line

braem commented:
Did those & went ahead and renamed & moved the file.

Hat-Kid commented on mp/game/momentum/scripts/credits.txt:
```diff
@@ -53,33 +53,48 @@
-        "Website Programming" "CreditsOutroTextLarge"
-        " " "CreditsOutroText"
+        "Larry" "CreditsOutroText"
+        "IDLPRUT" "CreditsOutroText"
```
ILDPRUT's name is misspelled here

Hat-Kid commented on mp/game/momentum/scripts/credits.txt:
```diff
@@ -53,33 +53,48 @@
+        "xen" "CreditsOutroText"
+        "Gouzi" "CreditsOutroText"
+        "rio" "CreditsOutroText"
+        "Hat-Kid" "CreditsOutroText"
```
I'd prefer if my name was just "Hat Kid"

braem commented:
- Moved the credits update to a different branch
- Removed border
- Fixed credits music to use music volume (uses both music and game volume)
- Hid cursor
